File: java/basic-async-bot/src/main/java/com/google/chat/bot/basic/async/Bot.java
/**
 * Copyright 2017 Google Inc.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.chat.bot.basic.async;

// [START async-bot]


import com.fasterxml.jackson.databind.JsonNode;
import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
import com.google.api.client.http.HttpRequestInitializer;
import com.google.api.client.json.jackson2.JacksonFactory;
import com.google.api.services.chat.v1.HangoutsChat;
import com.google.api.services.chat.v1.model.Message;
import com.google.api.services.chat.v1.model.Thread;
import com.google.auth.http.HttpCredentialsAdapter;
import com.google.auth.oauth2.GoogleCredentials;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import java.io.IOException;
import java.security.GeneralSecurityException;
import java.util.logging.Logger;

@SpringBootApplication
@RestController
public class Bot {
    static final String CHAT_SCOPE = "https://www.googleapis.com/auth/chat.bot";
    private static final Logger logger = Logger.getLogger(Bot.class.getName());

    public static void main(String[] args) {
        SpringApplication.run(Bot.class, args);
    }

    /**
     * Handles a GET request to the /bot endpoint.
     *
     * @param event Event from chat.
     * @return Message
     */
    @PostMapping("/")
    public void onEvent(@RequestBody JsonNode event) throws IOException, GeneralSecurityException {
        String replyText = "";
        switch (event.at("/type").asText()) {
            case "ADDED_TO_SPACE":
                String spaceType = event.at("/space/type").asText();
                if ("ROOM".equals(spaceType)) {
                    String displayName = event.at("/space/displayName").asText();
                    replyText = String.format("Thanks for adding me to %s", displayName);
                } else {
                    String displayName = event.at("/user/displayName").asText();
                    replyText = String.format("Thanks for adding me to a DM, %s!", displayName);
                }
                break;
            case "MESSAGE":
                String message = event.at("/message/text").asText();
                replyText = String.format("Your message: %s", message);
                break;
            case "REMOVED_FROM_SPACE":
                String name = event.at("/space/name").asText();
                logger.info(String.format("Bot removed from %s", name));
                break;
            default:
                replyText = "Cannot determine event type";
                break;
        }

        // [START async-response]
        String spaceName = event.at("/space/name").asText();
        Message reply = new Message().setText(replyText);
        // If replying to a message, set thread name to keep conversation together
        if (event.has("message")) {
            String threadName = event.at("/message/thread/name").asText();
            Thread thread = new Thread().setName(threadName);
            reply.setThread(thread);
        }

        GoogleCredentials credentials = GoogleCredentials.fromStream(
                Bot.class.getResourceAsStream("/service-acct.json")
        ).createScoped(CHAT_SCOPE);
        HttpRequestInitializer requestInitializer = new HttpCredentialsAdapter(credentials);
        HangoutsChat chatService = new HangoutsChat.Builder(
                    GoogleNetHttpTransport.newTrustedTransport(),
                    JacksonFactory.getDefaultInstance(),
                    requestInitializer)
                .setApplicationName("basic-async-bot-java")
                .build();
        chatService.spaces().messages().create(spaceName, reply).execute();
        // [END async-response]
    }
}

// [END async-bot]

File: python/pub-sub-bot/bot.py
# Copyright 2017 Google Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""
Hangouts Chat bot that listens for messages via Cloud Pub/Sub.
"""

# [START pub-sub-bot]

import json
import logging
import os
import sys
import time
from google.cloud import pubsub_v1
from googleapiclient.discovery import build
import google.auth


def receive_messages():
  """Receives messages from a pull subscription."""

  scopes = ['https://www.googleapis.com/auth/chat.bot']
  credentials, project_id = google.auth.default()
  credentials = credentials.with_scopes(scopes=scopes)
  chat = build('chat', 'v1', credentials=credentials)

  subscription_id = os.environ.get('SUBSCRIPTION_ID')
  subscriber = pubsub_v1.SubscriberClient()
  subscription_path = subscriber.subscription_path(
      project_id, subscription_id)

  def callback(message):
    logging.info('Received message: %s', message.data)

    event = json.loads(message.data)
    space_name = event['space']['name']

    # If the bot was removed, we don't need to return a response.
    if event['type'] == 'REMOVED_FROM_SPACE':
      logging.info('Bot removed rom space %s', space_name)
      return

    response = format_response(event)

    # Send the asynchronous response back to Hangouts Chat
    chat.spaces().messages().create(
        parent=space_name,
        body=response).execute()
    message.ack()

  subscriber.subscribe(subscription_path, callback=callback)
  logging.info('Listening for messages on %s', subscription_path)

  # Keep main thread from exiting while waiting for messages
  while True:
    time.sleep(60)


def format_response(event):
  """Determine what response to provide based upon event data.
  Args:
    event: A dictionary with the event data.
  """

  event_type = event['type']

  text = ""
  sender_name = event['user']['displayName']

  match event_type:
    case 'ADDED_TO_SPACE':
      if event['space']['type'] == 'ROOM':
        text = 'Thanks for adding me to {}!'.format(
            event['space']['displayName'])
      else:
        text = 'Thanks for adding me to a DM, {}!'.format(sender_name)

    case 'MESSAGE':
      # The following three lines of code update the thread that raised the event.
      # Remove them if you want to send the message in a new thread.
      if event['message']['thread'] is not None:
        thread_id = event['message']['thread']
        response['thread'] = thread_id

      text = 'Your message, {}: "{}"'.format(
          sender_name, event['message']['text'])

  response = {'text': text}

  return response


if __name__ == '__main__':
  if 'SUBSCRIPTION_ID' not in os.environ:
    logging.error('Missing SUBSCRIPTION_ID env var.')
    sys.exit(1)

  logging.basicConfig(
      level=logging.INFO,
      style='{',
      format='{levelname:.1}{asctime} {filename}:{lineno}] {message}')
  receive_messages()

# [END pub-sub-bot]
